### PR TITLE
⚠️ Switch to 1 secret per certificate

### LINKED ISSUE #135: Modify cert secret to be 1 secret per cert
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
In  https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/pull/126 certificate generation and secret lookup was added to the bootstrap controller. We had a discussion on the secret format related to cert-manager. In order to support cert-manager we will need 1 secret per cert. 

See details here: https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/pull/126#discussion_r314039485

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by moshloop**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** 
Fixes #135 

**Special notes for your reviewer**:
Still need to cleanup a little, but wanted early feedback. 

One of the outstanding items I think is to create an enum for the certificate names - the question is if that should go in `api` or in the packages?

/assign @ncdc @chuckha

### CODE REVIEW

File: certs/certs.go [hunk 1]
```diff
@@ -34,6 +35,11 @@ const (
 	duration365d = time.Hour * 24 * 365
 )
 
+// ClusterCertificatesSecretName returns the name of the certificates secret, given a cluster name
+func ClusterCertificatesSecretName(clusterName string) string {
+	return fmt.Sprintf("%s-certs", clusterName)
```
Comment (ncdc) [inline certs/certs.go]: This is not unique to this PR, and we have other places in master that have this issue, but I wanted to point it out: if the clusterName is already at the max length, then clusterName-certs will be too long and the apiserver will reject it.

File: certs/certs.go [hunk 2]
```diff
@@ -42,6 +48,38 @@ type Certificates struct {
 	ServiceAccount *KeyPair
 }
 
+// Set sets a certificate by name, ignoring unknown names
+func (certs *Certificates) Set(name string, keyPair *KeyPair) {
+	if name == "ca" {
```
Comment (ncdc) [inline certs/certs.go]: Can you use a `switch` statement instead, and also use constants for the names?

File: certs/certs.go [hunk 3]
```diff
@@ -42,6 +48,38 @@ type Certificates struct {
...
+		return certs.EtcdCA
+	} else if name == "sa" {
+		return certs.ServiceAccount
+	} else if name == "proxy" {
+		return certs.FrontProxyCA
+	}
+	return nil
+}
+
+var ClusterCaName = "ca"
```
Comment (ncdc) [inline certs/certs.go]: I'm guessing you meant for these to be `const`s?

File: certs/secrets.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
+package certs
```
Comment (ncdc) [inline certs/secrets.go]: Please add the boilerplate header

File: certs/secrets.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+import (
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (ncdc) [inline certs/secrets.go]: We usually alias this to `metav1`

File: certs/secrets.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	"strings"
+)
+
+// NewSecretsFromCertificates returns a list of new secrets, 1 for each certificate
+func NewSecretsFromCertificates(input *Certificates) []*corev1.Secret {
```
Comment (ncdc) [inline certs/secrets.go]: Do you want to check for each input keypair and return an error if any is missing (like NewCertificatesFromSecrets)?

File: certs/secrets.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+// NewCertificatesFromSecrets returns the certificates for the cluster by retrieving
+// each certificate from 1 secret
+func NewCertificatesFromSecrets(secrets *corev1.SecretList) (*Certificates, error) {
+	var certs Certificates
+	for _, secret := range secrets.Items {
+		certs.Set(getCertName(secret.ObjectMeta.Name), getKeyPair(secret))
+	}
+	if certs.ClusterCA == nil {
+		return nil, errors.Errorf("Unable to find secret for ClusterCA, got secrets: %s", getSecretNames(secrets))
```
Comment (ncdc) [inline certs/secrets.go]: Error messages need to start with a lowercase letter

File: certs/secrets.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	var names []string
+	for _, secret := range secrets.Items {
+		names = append(names, secret.ObjectMeta.Name)
+	}
+	return strings.Join(names, ", ")
+}
+
+func getKeyPair(secret corev1.Secret) *KeyPair {
+	return &KeyPair{
+		Cert: secret.Data["tls.crt"],
```
Comment (ncdc) [inline certs/secrets.go]: Could you make `"tls.crt"` and `"tls.key"` constants?

File: certs/utils.go [hunk 9]
```diff
@@ -78,6 +79,16 @@ func NewCertificatesFromMap(m map[string][]byte) *Certificates {
 	return certs
 }
 
+// // ToMap converts Certificates into a map
```
Comment (ncdc) [inline certs/utils.go]: ?

File: controllers/kubeadmconfig_controller.go [hunk 10]
```diff
@@ -449,30 +455,40 @@ func (r *KubeadmConfigReconciler) createClusterCertificates(ctx context.Context,
...
-					Kind:       "KubeadmConfig",
-					Name:       config.GetName(),
-					UID:        config.GetUID(),
-				},
-			},
-		},
-		Data: certificates.ToMap(),
+	for _, secret := range certs.NewSecretsFromCertificates(certificates) {
+		linkOwnerReference(&secret.ObjectMeta, config)
+		secret.ObjectMeta.Labels["cluster"] = clusterName
```
Comment (ncdc) [inline controllers/kubeadmconfig_controller.go]: Let's use the same label we already have - `"cluster.x-k8s.io/cluster-name"`. We have a constant for this, `MachineClusterLabelName`, in `sigs.k8s.io/cluster-api/api/v1alpha2`. We can either use this, or create our own constant in this repo.

File: controllers/kubeadmconfig_controller.go [hunk 11]
```diff
@@ -449,30 +455,40 @@ func (r *KubeadmConfigReconciler) createClusterCertificates(ctx context.Context,
...
+			Kind:       "KubeadmConfig",
+			Name:       config.GetName(),
+			UID:        config.GetUID(),
+		},
 	}
+}
 
-	return certificates, nil
+func prefixByClusterName(clusterName, name string) string {
+	return fmt.Sprintf("%s-%s", clusterName, name)
```
Comment (ncdc) [inline controllers/kubeadmconfig_controller.go]: Same comment as above about possibly exceeding max name length restrictions

File: controllers/kubeadmconfig_controller.go [hunk 12]
```diff
@@ -449,30 +455,40 @@ func (r *KubeadmConfigReconciler) createClusterCertificates(ctx context.Context,
...
+			return nil, err
+		}
 	}
 
-	if err := r.Create(ctx, secret); err != nil {
-		return nil, err
+	return certificates, nil
+}
+
+func linkOwnerReference(meta *v1.ObjectMeta, config *cabpkv1alpha2.KubeadmConfig) {
```
Comment (ncdc) [inline controllers/kubeadmconfig_controller.go]: I'd rather see this function generate and return the ownerReferences, and set the secret's ownerReferences in the for loop above using the output from this function.

File: certs/certs.go [hunk 13]
```diff
@@ -30,8 +30,18 @@ import (
...
+	tlsKey       = "tls.key"
+	tlsCrt       = "tls.crt"
+	//ClusterCaName is the secret name suffix for apiserver CA
+	ClusterCaName = "ca"
+	//EtcdCaName is the secret name suffix for the Etcd CA
+	EtcdCaName = "etcd"
+	//ServiceAccountName is the secret name suffix for the Service Account keys
+	ServiceAccountName = "sa"
+	//FrontProxyName is the secret name suffix for Front Proxy CA
+	FrontProxyName = "proxy"
```
Comment (detiber) [inline certs/certs.go]: Should this be `FrontProxyCaName` to go with the naming convention of the other CA vars?

File: certs/certs.go [hunk 14]
```diff
@@ -30,8 +30,18 @@ import (
 )
 
 const (
-	rsaKeySize   = 2048
 	duration365d = time.Hour * 24 * 365
+	rsaKeySize   = 2048
+	tlsKey       = "tls.key"
+	tlsCrt       = "tls.crt"
+	//ClusterCaName is the secret name suffix for apiserver CA
+	ClusterCaName = "ca"
```
Comment (chuckha) [inline certs/certs.go]: `Ca` should be CA as it's an acronym

**Comment by moshloop** [comment]
@chuckha yes I will update PR in a day or 2

**Comment by chuckha** [comment]
I successfully got a multi-control plane cluster up and running with this branch! Thanks so much for putting it together. LGTM, just need to address @detiber's feedback 👍

**Comment by moshloop** [comment]
I will update the PR by Sunday night latest - I also want to make specifying the etcd,proxy and sa certs optional

**Comment by ncdc** [comment]
Superseded by #197
/close

**Comment by ncdc** [comment]
@moshloop the team told me you were cool with one of us taking your PR and carrying it forward today.

**Pull request closed without merging**